src/Cli/dotnet/Program.cs

```diff
@@ -127,10 +128,23 @@ internal static int ProcessArgs(string[] args, TimeSpan startupTime)
         ParseResult parseResult;
         using (new PerformanceMeasurement(performanceData, "Parse Time"))
         {
-            // If we get C# file path as the first argument, parse as `dotnet run file.cs`.
-            parseResult = args is [{ } filePath, ..] && VirtualProjectBuildingCommand.IsValidEntryPointPath(filePath)
-                ? Parser.Instance.Parse(["run", .. args])
-                : Parser.Instance.Parse(args);
+            parseResult = Parser.Instance.Parse(args);
+            // If we get didn't match any built-in commands, and a C# file path is the first argument,
+            // parse as `dotnet run file.cs ..rest_of_args` instead.
+            if (parseResult.CommandResult.Command is RootCommand
+                && parseResult.GetValue(Parser.DotnetSubCommand) is { } unmatchedCommandOrFile
```

Comment:
This `DotnetSubCommand` argument (misnomer alert!) is also used later in the Program.cs when we try to run globally-installed tools with the naming convention `dotnet-<arg>`. It's intended for analysis like this!